File: Tagger/POS_Tagger.js
import fs from 'fs';

function tokenize(text) {
  const abbreviations = new Set([
    "Mr.",
    "Dr.",
    "U.S.",
    "Ph.D.",
    "e.g.",
    "i.e.",
  ]);
  let words = text.match(/\b[\w'.-]+\b/g) || [];

  return words.map((word) => {
    if (!abbreviations.has(word) && /[.,!?]$/.test(word)) {
      return word.slice(0, -1);
    }
    return word;
  });
}

function determinePOSTag(word, prevWord) {
  if (/.*ed$/.test(word)) return "VBD"; 
  if (/.*ing$/.test(word)) return "VBG"; 
  if (/.*s$/.test(word)) return "NNS"; 

  if (["a", "an", "the"].includes(prevWord)) return "NN"; 
  if (/^[A-Z]/.test(word) && prevWord) return "NNP"; 
  if (["is", "was", "are"].includes(word)) return "VBZ"; 

  return "NN"; 
}

fs.readFile("D:/WEB DEVLOPMENT/NODE_JS/Tagger/dictionary.txt", "utf8", (err, data) => {
  if (err) {
    console.error("Error opening file!");
    return;
  }

  let words = tokenize(data);
  let taggedWords = [];
  let prevWord = "";

  words.forEach((word) => {
    let posTag = determinePOSTag(word, prevWord);
    taggedWords.push(`${word}/${posTag}`);
    prevWord = word;
  });

  console.log("POS Tagged Text:");
  console.log(taggedWords.join(" "));
});

File: Tagger/POS_Tagger.cpp
// POS Tagger 
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_set>
#include <vector>
#include <cctype>
#include <regex>

using namespace std;


vector<string> tokenize(const string &text) {
    vector<string> words;
    stringstream ss(text);
    string word;

    unordered_set<string> abbreviations = {"Mr.", "Dr.", "U.S.", "Ph.D.", "e.g.", "i.e."};
    while (ss >> word) {

        if (!abbreviations.count(word) && ispunct(word.back())) {
            word.pop_back();
        }
        words.push_back(word);
    }

    return words;
}


string determinePOSTag(const string &word, const string &prevWord) {

    regex pastTenseVerb(".*ed$");   
    regex gerundVerb(".*ing$");     
    regex pluralNoun(".*s$");       
    
    if (regex_match(word, pastTenseVerb)) return "VBD";  
    if (regex_match(word, gerundVerb)) return "VBG";     
    if (regex_match(word, pluralNoun)) return "NNS";    
    
    if (prevWord == "a" || prevWord == "an" || prevWord == "the") return "NN";

    if (isupper(word[0]) && prevWord != "") return "NNP";

    if (word == "is" || word == "was" || word == "are") return "VBZ";

    return "NN";  
}

int main() {
    ifstream file("dictionary.txt");
    if (!file) {
        cerr << "Error opening file!" << endl;
        return 1;
    }

    string text, line;
    while (getline(file, line)) {
        text += line + " ";
    }
    file.close();

    vector<string> words = tokenize(text);
    vector<pair<string, string>> taggedWords;
    
    string prevWord = "";
    for (const string &word : words) {
        string posTag = determinePOSTag(word, prevWord);
        taggedWords.push_back({word, posTag});
        prevWord = word;  
    }

    cout << "POS Tagged Text:\n";
    for (const auto &tagged : taggedWords) {
        cout << tagged.first << "/" << tagged.second << " ";
    }
    cout << endl;

    return 0;
}
